pyo3-macros-backend/src/pyproto.rs

```diff
@@ -129,49 +130,61 @@ fn submit_normal_methods(py_methods: Vec<TokenStream>, ty: &syn::Type) -> TokenS
     }
 }
 
-fn submit_protocol_methods(
+fn impl_proto_methods(
     method_names: HashSet<String>,
     ty: &syn::Type,
     proto: &defs::Proto,
-) -> syn::Result<TokenStream> {
-    if proto.extension_trait == "" {
-        return Ok(quote! {});
+) -> TokenStream {
+    if proto.slots_trait.is_empty() {
+        return TokenStream::default();
     }
-    let ext_trait: syn::Path = syn::parse_str(proto.extension_trait)?;
-    let mut tokens = vec![];
+
+    let module = proto.module();
+    let slots_trait = syn::Ident::new(proto.slots_trait, Span::call_site());
+    let slots_trait_slots = syn::Ident::new(proto.slots_trait_slots, Span::call_site());
+
     if proto.name == "Buffer" {
-        // For buffer, we construct `PyProtoMethods` from PyBufferProcs
-        tokens.push(quote! {
-            let mut proto_methods = pyo3::ffi::PyBufferProcs::default();
-        });
-        for getter in proto.slot_getters(method_names) {
-            let get = syn::Ident::new(getter, Span::call_site());
-            let field = syn::Ident::new(&format!("bf_{}", &getter[4..]), Span::call_site());
-            tokens.push(quote! { proto_methods.#field = Some(<#ty as #ext_trait>::#get()); });
-        }
-    } else {
-        // For other protocols, we construct `PyProtoMethods` from Vec<ffi::PyType_Slot>
-        tokens.push(quote! { let mut proto_methods = vec![]; });
-        for getter in proto.slot_getters(method_names) {
-            let get = syn::Ident::new(getter, Span::call_site());
-            tokens.push(quote! {
-                let slot = <#ty as #ext_trait>::#get();
-                proto_methods.push(pyo3::ffi::PyType_Slot { slot: slot.0, pfunc: slot.1 as _ });
-            });
-        }
-    };
-    if tokens.len() <= 1 {
-        return Ok(quote! {});
+        return quote! {
+            impl pyo3::class::proto_methods::#slots_trait<#ty>
+            for pyo3::class::proto_methods::PyClassProtocols<#ty> {
```

Comment:
Just a nit, but how about more `rustfmt`ic?
```suggestion
            impl pyo3::class::proto_methods::#slots_trait<#ty>
                for pyo3::class::proto_methods::PyClassProtocols<#ty> 
            {
```